container/podman/factory.go

```diff
@@ -0,0 +1,211 @@
+	cgroupSubsystems libcontainer.CgroupSubsystems
+
+	// Information about mounted filesystems.
+	fsInfo fs.FsInfo
+
+	podmanVersion []int
+
+	podmanAPIVersion []int
+
+	includedMetrics container.MetricSet
+}
+
+func (f *podmanFactory) String() string {
+	return PodmanNamespace
+}
+
+func (f *podmanFactory) NewContainerHandler(name string, inHostNamespace bool) (handler container.ContainerHandler, err error) {
+	client, err := Client()
+	if err != nil {
+		return
+	}
+
+	metadataEnvs := strings.Split(*podmanEnvWhitelist, ",")
+
+	handler, err = newPodmanContainerHandler(
+		client,
+		name,
+		f.machineInfoFactory,
+		f.fsInfo,
+		&f.cgroupSubsystems,
+		inHostNamespace,
+		metadataEnvs,
+		f.podmanVersion,
+		f.includedMetrics,
+	)
+	return
+}
+
+// Returns the Podman ID from the full container name.
+func CgroupNameToPodmanId(name string) string {
+	id := path.Base(name)
+
+	if matches := podmanCgroupRegexp.FindStringSubmatch(id); matches != nil {
+		return matches[1]
+	}
+
+	return id
+}
+
+// isContainerName returns true if the cgroup with associated name
+// coult be a podman container.
+// the actual decision is made by running a ContainerInspect API call
+func isContainerName(name string) bool {
+	// always ignore .mount cgroup even if associated with podman and delegate to systemd
+	if strings.HasSuffix(name, ".mount") {
+		return false
+	}
+	return podmanCgroupRegexp.MatchString(path.Base(name))
+}
+
+// Podman handles all containers prefixed with libpod-
+func (f *podmanFactory) CanHandleAndAccept(name string) (bool, bool, error) {
+	// if the container is not associated with podman, we can't handle it or accept it.
+	if !isContainerName(name) {
+		return false, false, nil
+	}
+
+	// Check if the container is known to podman and it is active.
+	id := CgroupNameToPodmanId(name)
+
+	// We assume that if Inspect fails then the container is not known to podman.
+	ctnr, err := f.client.ContainerInspect(context.Background(), id)
+	if err != nil || !ctnr.State.Running {
+		return false, true, fmt.Errorf("error inspecting container: %v", err)
+	}
+
+	return true, true, nil
+}
+
+func (f *podmanFactory) DebugInfo() map[string][]string {
+	return map[string][]string{}
+}
+
+var (
+	versionRegexpString    = `(\d+)\.(\d+)\.(\d+)`
+	versionRe              = regexp.MustCompile(versionRegexpString)
+	apiVersionRegexpString = `(\d+)\.(\d+)`
+	apiVersionRe           = regexp.MustCompile(apiVersionRegexpString)
+)
+
+// Register root container before running this function!
+func Register(factory info.MachineInfoFactory, fsInfo fs.FsInfo, includedMetrics container.MetricSet) error {
+	client, err := Client()
+	if err != nil {
+		return fmt.Errorf("unable to communicate with podman: %v", err)
+	}
+
+	podmanInfo, err := ValidateInfo()
+	if err != nil {
+		return fmt.Errorf("failed to validate Podman info: %v", err)
+	}
+
+	// Version already validated above, assume no error here.
+	podmanVersion, _ := parseVersion(podmanInfo.ServerVersion, versionRe, 3)
```

Comment:
If the version is already validated then why don't you just split a string? Regular expression might be an overkill in this scenario.